Integrations/Snowflake/Snowflake.py

```diff
@@ -0,0 +1,392 @@
+    if isinstance(results, dict):
+        results = process_table_row(results, checks)
+    else:
+        # if 'results' isn't a dict, assume it's a list
+        for i, row in enumerate(results):
+            results[i] = process_table_row(row, checks)
+    return results
+
+
+def get_connection(args):
+    """
+    Build the connection based on parameters
+
+    parameter: (dict) args
+        The command arguments of the command function calling this helper function
+
+    returns:
+        Snowflake connection
+    """
+    params = {}
+    set_provided(params, 'user', USER)
+    set_provided(params, 'password', PASSWORD)
+    set_provided(params, 'account', ACCOUNT)
+    set_provided(params, 'authenticator', AUTHENTICATOR)
+    set_provided(params, 'region', REGION)
+    set_provided(params, 'warehouse', args.get('warehouse'), WAREHOUSE)
+    set_provided(params, 'database', args.get('database'), DATABASE)
+    set_provided(params, 'schema', args.get('schema'), SCHEMA)
+    set_provided(params, 'role', args.get('role'), ROLE)
+    if CERTIFICATE:
+        p_key = serialization.load_pem_private_key(CERTIFICATE, backend=default_backend())
+        pkb = p_key.private_bytes(
+            encoding=serialization.Encoding.DER,
+            format=serialization.PrivateFormat.PKCS8,
+            encryption_algorithm=serialization.NoEncryption())
+        params['private_key'] = pkb
+    return snowflake.connector.connect(**params)
+
+
+def row_to_incident(column_descriptions, row):
+    """
+    Create incident from data returned by queried database in fetch_incidents
+
+    parameter: (list) column_descriptions
+        The metadata that describes the values for each column in the 'data' parameter
+
+    parameter: (dict) row
+        The row of data where each cell's key in the row is the name of the column
+        to which it belongs
+
+    returns:
+        Incident Object
+    """
+    incident = {}
+    occurred = row.get(DATETIME_COLUMN)
+    timestamp = None
+    if occurred:
+        if isinstance(occurred, dttime) or isinstance(occurred, timedelta):
+            err_msg = 'The datetime field specified in the integration parameters must '
+            err_msg += 'contain values of type "datetime" or "date".'
+            raise FetchError(err_msg)
+        timestamp = occurred.timestamp() * 1000
+    else:
+        err_msg = 'Nothing found when trying to fetch the datetime field specified in'
+        err_msg += ' the integration parameters. Please check that the name was correct.'
+        err_msg += ' If the field name was correct, verify that the returned value for'
+        err_msg += ' the specified field is not NULL for ALL of the rows to be fetched.'
+        raise FetchError(err_msg)
+    # Incident Title
+    if INCIDENT_NAME_COLUMN:
+        name = row.get(INCIDENT_NAME_COLUMN)
+    else:
+        name = 'Snowflake Incident -- ' + convert_datetime_to_string(occurred) + '-' + datetime.now().timestamp()
```

Comment:
why do we need two timestamps here? isnt there some unique key to use instead of the timestamps ?

Reply:
I needed a way to make incident names unique in the case that the user configuring the instance of the integration didn't supply the name of a table column whose value would be used in assigning the incident name (Maya's advice was to not make it required)